helix-core/src/syntax.rs

```diff
@@ -712,6 +718,34 @@ thread_local! {
     })
 }
 
+/// Creates an iterator over the captures in a query within the given range,
+/// re-using a cursor from the pool if available.
+fn query_captures<'a, 'tree>(
```

Comment:
```suggestion
/// Safety: The `QueryCaptures` must bed dropped before the `QueryCusor` is dropped
unsafe fn query_captures<'a, 'tree>(
```

This function signature makes me very uncomfortable because it can easily lead to use after free, alias violations and other kinds of UB. The way it is used currently is fine but it should be marked unsafe so it doesn't get used incorrectly in the future.

Really the crux of this function is that you need an owning `QueryCaptures` iterator that owns the `QueryCursor` (and free it on `Drop`) and can be turned back into a `QueryCursor`. That kind of API could be implemented here but it would be a somewhat ugly hack. This should really be supported inside the tree sitter bindings directly. I will be submitting a breaking change to that API anyway and try to include an appropriate type with that change.

For now marking this unsafe should be good enough tough so we don't block this PR on the TS changes